### ISSUE TITLE
`st_wrap` incorrectly assigns values for non overlapping grid when having irregular source

### ISSUE THREAD

**Issue by bart1** [r-spatial/stars#481]
I noticed `st_warp` incorrectly assigns values to grid that is non overlapping with the original grid when the source grid is irregular. In the example below this is `r` while this works correctly for `r2`. This can be seen as the first image returns values while the second an empty plot as is correct.

``` r
require(stars)
#> Loading required package: stars
#> Loading required package: abind
#> Loading required package: sf
#> Linking to GEOS 3.8.0, GDAL 3.0.4, PROJ 6.3.1
r <- st_as_stars(list(values = matrix(1:20, ncol = 4)),
  dimensions = st_dimensions(
    x = c(0, 0.5, 1, 2, 4, 5),
    y = c(0.3, 0.5, 1, 2, 2.2)
  )
)
sf::st_crs(r) <- 4326
r2 <- st_as_stars(list(values = matrix(1:20, ncol = 4)),
  dimensions = st_dimensions(
    x = 1:6,
    y = 1:5
  )
)
sf::st_crs(r2) <- 4326
grid <- structure(list(values = structure(c(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0), .Dim = c(x = 4L, y = 4L))), dimensions = structure(list(x = structure(list(from = 1, to = 4L, offset = -1252344.27142433, delta = 156543.033928041, refsys = structure(list(input = "EPSG:3857", wkt = "PROJCRS[\"WGS 84 / Pseudo-Mercator\",\n    BASEGEOGCRS[\"WGS 84\",\n        DATUM[\"World Geodetic System 1984\",\n            ELLIPSOID[\"WGS 84\",6378137,298.257223563,\n                LENGTHUNIT[\"metre\",1]]],\n        PRIMEM[\"Greenwich\",0,\n            ANGLEUNIT[\"degree\",0.0174532925199433]],\n        ID[\"EPSG\",4326]],\n    CONVERSION[\"Popular Visualisation Pseudo-Mercator\",\n        METHOD[\"Popular Visualisation Pseudo Mercator\",\n            ID[\"EPSG\",1024]],\n        PARAMETER[\"Latitude of natural origin\",0,\n            ANGLEUNIT[\"degree\",0.0174532925199433],\n            ID[\"EPSG\",8801]],\n        PARAMETER[\"Longitude of natural origin\",0,\n            ANGLEUNIT[\"degree\",0.0174532925199433],\n            ID[\"EPSG\",8802]],\n        PARAMETER[\"False easting\",0,\n            LENGTHUNIT[\"metre\",1],\n            ID[\"EPSG\",8806]],\n        PARAMETER[\"False northing\",0,\n            LENGTHUNIT[\"metre\",1],\n            ID[\"EPSG\",8807]]],\n    CS[Cartesian,2],\n        AXIS[\"easting (X)\",east,\n            ORDER[1],\n            LENGTHUNIT[\"metre\",1]],\n        AXIS[\"northing (Y)\",north,\n            ORDER[2],\n            LENGTHUNIT[\"metre\",1]],\n    USAGE[\n        SCOPE[\"unknown\"],\n        AREA[\"World - 85°S to 85°N\"],\n        BBOX[-85.06,-180,85.06,180]],\n    ID[\"EPSG\",3857]]"), class = "crs"), point = NA, values = NULL), class = "dimension"), y = structure(list(from = 1, to = 4L, offset = 626172.135712164, delta = -156543.033928041, refsys = structure(list(input = "EPSG:3857", wkt = "PROJCRS[\"WGS 84 / Pseudo-Mercator\",\n    BASEGEOGCRS[\"WGS 84\",\n        DATUM[\"World Geodetic System 1984\",\n            ELLIPSOID[\"WGS 84\",6378137,298.257223563,\n                LENGTHUNIT[\"metre\",1]]],\n        PRIMEM[\"Greenwich\",0,\n            ANGLEUNIT[\"degree\",0.0174532925199433]],\n        ID[\"EPSG\",4326]],\n    CONVERSION[\"Popular Visualisation Pseudo-Mercator\",\n        METHOD[\"Popular Visualisation Pseudo Mercator\",\n            ID[\"EPSG\",1024]],\n        PARAMETER[\"Latitude of natural origin\",0,\n            ANGLEUNIT[\"degree\",0.0174532925199433],\n            ID[\"EPSG\",8801]],\n        PARAMETER[\"Longitude of natural origin\",0,\n            ANGLEUNIT[\"degree\",0.0174532925199433],\n            ID[\"EPSG\",8802]],\n        PARAMETER[\"False easting\",0,\n            LENGTHUNIT[\"metre\",1],\n            ID[\"EPSG\",8806]],\n        PARAMETER[\"False northing\",0,\n            LENGTHUNIT[\"metre\",1],\n            ID[\"EPSG\",8807]]],\n    CS[Cartesian,2],\n        AXIS[\"easting (X)\",east,\n            ORDER[1],\n            LENGTHUNIT[\"metre\",1]],\n        AXIS[\"northing (Y)\",north,\n            ORDER[2],\n            LENGTHUNIT[\"metre\",1]],\n    USAGE[\n        SCOPE[\"unknown\"],\n        AREA[\"World - 85°S to 85°N\"],\n        BBOX[-85.06,-180,85.06,180]],\n    ID[\"EPSG\",3857]]"), class = "crs"), point = NA, values = NULL), class = "dimension")), raster = structure(list(affine = c(0, 0), dimensions = c("x", "y"), curvilinear = FALSE), class = "stars_raster"), class = "dimensions"), class = "stars")
grid
#> stars object with 2 dimensions and 1 attribute
#> attribute(s):
#>         Min. 1st Qu. Median Mean 3rd Qu. Max.
#> values     0       0      0    0       0    0
#> dimension(s):
#>   from to   offset   delta                   refsys point values x/y
#> x    1  4 -1252344  156543 WGS 84 / Pseudo-Mercator    NA   NULL [x]
#> y    1  4   626172 -156543 WGS 84 / Pseudo-Mercator    NA   NULL [y]
image(st_warp(r, grid))
```

![](https://i.imgur.com/VYX2Kps.png)

``` r
image(st_warp(r2, grid))
#> Warning in min(x): no non-missing arguments to min; returning Inf
#> Warning in max(x): no non-missing arguments to max; returning -Inf
```

![](https://i.imgur.com/bxPPAob.png)

``` r
require(ggplot2)
#> Loading required package: ggplot2
ggplot() +
  geom_stars(data = r, sf = T) +
  geom_stars(data = grid, sf = T)
```

![](https://i.imgur.com/HqBuFxc.png)

``` r
ggplot() +
  geom_stars(data = r2, sf = T) +
  geom_stars(data = grid, sf = T)
```

![](https://i.imgur.com/qRQTHqk.png)

<sup>Created on 2021-11-30 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by edzer** [comment]
This should fix it, but creates a dependency on R >= 4.1.0 (where `apply` gained the `simplify` argument).

**Comment by bart1** [comment]
Thanks it works!
